tests/test_compare_dictconfig_vs_dict.py

```diff
@@ -0,0 +1,211 @@
+"""
+This file compares dunder methods of DictConfig (e.g. __getitem__ and
+__setitem__) with dunder methods of standard python's dict.
+
+We have separate test classes for the following cases:
+    TestUntypedDictConfig: for DictConfig without a set key_type
+    TestPrimitiveTypeDunderMethods: for DictConfig where key_type is primitive
+    TestEnumTypeDunderMethods: for DictConfig where key_type is Enum
+"""
+from enum import Enum
+from typing import Any, Dict
+
+import pytest
+
+from omegaconf import DictConfig
+from omegaconf.errors import KeyValidationError
+from tests import Enum1
+
+
+@pytest.fixture(
+    params=[
+        "a",
+        "b",
+        1,
+        2,
+        123.45,
+        67.89,
+        True,
+        False,
+        Enum1.FOO,
+        Enum1.BAR,
+        "FOO",
+        "BAR",
+    ]
+)
+def key(request: Any) -> Any:
+    """A key to test indexing into DictConfig."""
+    return request.param
```

Comment:
Not familiar with this style of fixtures. What does it do?

Reply:
It is very similar to `@pytest.parametrize`.
When a test function uses the `key` fixture, the test will be parametrized so that each value in the `params` list gets used as a `key`.

Reply:
I see.
So this is an alternative form to :
```python
@pytest.mark.parameterize("a", [...])
class TestFoo:
  def test_bar(a):
    ...

  def test_rab(a):
    ...
```
?

Reply:
Yes, exactly.
One possible benefit is that you can use one fixture as input to another fixture, so you could for example write
```python
@pytest.fixture(params=["key_a","key_b"])
def key(request: Any) -> str:
    return request.param

@pytest.fixture(params=["lock1", "lock2"])
def key_in_lock(key, request: Any) -> str:
    lock = request.param
    return lock + "(" + key + ")"

def test_key_in_lock(key_in_lock):
    print(key_in_lock)
```
which outputs
```
lock1(key_a)
lock1(key_b)
lock2(key_a)
lock2(key_b)
```

Reply:
I started to experiment with this in the structured_ir PR.

Reply:
:) Nice!